### ISSUE TITLE
lower priority job can not be preempted by high priority job when lower priority job enable gang

### ISSUE THREAD

**Issue by nkflash** [volcano-sh/volcano#2321]
I have 2 user define queues and one default queue(default).

like:

low priority queue:

xbmin@p-ksyun-develop-001:/mnt/data/xbmin/kubeflow_test/config$ cat low.yaml
apiVersion: scheduling.volcano.sh/v1beta1
kind: Queue
metadata:
  name: low-queue
  namespace: kubeflow
spec:
  reclaimable: true
  weight: 1
  capability:
    cpu: "96"
    memory: "359Gi"
status:
  state: Open

high priority queue:
xbmin@p-ksyun-develop-001:/mnt/data/xbmin/kubeflow_test/config$ cat high.yaml
apiVersion: scheduling.volcano.sh/v1beta1
kind: Queue
metadata:
  name: high-queue
  namespace: kubeflow
spec:
  weight: 3
  reclaimable: true
  capability:
    cpu: "96"
    memory: "359Gi"
status:
  state: Open

There are only 96 core in cluster. And total queue weight is 5(low: 1, default: 1, high: 3)
I submit 2 lower priority jobs into low queue. Each job require 34 CPUs. So  total require 68 CPUs
Then I submit a higher job to high queue, which require 34 CPUs. Base on the "reclaim" action.
The high job should run and low jobs should be evicted, but I got the low job still occupy the resource. 
The high job can not schedule.

It seems the code forbid that. Only allocated resources count larger than MinAvailable job could be preempted.
![image](https://user-images.githubusercontent.com/42488903/176857434-b5053d8a-4171-4ce4-8412-15b3d557d292.png)

This is not reasonable, lower priority gang job also need to be preempted no matter if it satisfied by MinAvailable.
When I change that code, the reclaim work well.

**Comment by Thor-wl** [comment]
Hi, thanks for your feedback. I reviewed the scenario you mentioned above. As the original design, the following points are important for users.
* Preempt only works for jobs within the same queue if users configure `preempt` action. In most scenarios, different jobs come from different projects or teams. It is not so reasonable to destory others' jobs.
* Reclaim works between different queues. It may trigger resource preemption from other queues(other projects or teams). But the preemption action promises the least harmless and works without priority comparsion between jobs from different queues.

**Comment by nkflash** [comment]
> Hi, thanks for your feedback. I reviewed the scenario you mentioned above. As the original design, the following points are important for users.
> 
> * Preempt only works for jobs within the same queue if users configure `preempt` action. In most scenarios, different jobs come from different projects or teams. It is not so reasonable to destory others' jobs.
> * Reclaim works between different queues. It may trigger resource preemption from other queues(other projects or teams). But the preemption action promises the least harmless and works without priority comparsion between jobs from different queues.

In my scenario, queues's weight are different. I expect the overused queue should return the resources when the other queue have job coming. It is sure the priority is not the important point for this scenario. Without gang, it works as my expectation. The issue is if I enable gang, the overused resource queue cannot return the resources to the other queue.

**Comment by Thor-wl** [comment]
> > Hi, thanks for your feedback. I reviewed the scenario you mentioned above. As the original design, the following points are important for users.
> > 
> > * Preempt only works for jobs within the same queue if users configure `preempt` action. In most scenarios, different jobs come from different projects or teams. It is not so reasonable to destory others' jobs.
> > * Reclaim works between different queues. It may trigger resource preemption from other queues(other projects or teams). But the preemption action promises the least harmless and works without priority comparsion between jobs from different queues.
> 
> In my scenario, queues's weight are different. I expect the overused queue should return the resources when the other queue have job coming. It is sure the priority is not the important point for this scenario. Without gang, it works as my expectation. The issue is if I enable gang, the overused resource queue cannot return the resources to the other queue.

I got you. Just as what I mentioned above, all the preemption action works based on the condition that it is not harmful to the success of the business process. So it should be judged that scheduler cannot evicted pods in which job its pod number is less than `minAvailable`.

**Comment by nkflash** [comment]
> 

I see your point. But for AI workload, gang is by default scheduling policy(Since most AI frameworks don't support failover). In such situation, reclaim will not work at all. So I think "pod number is less than minAvailable" should support configure, let user to decide if gang job(lower priority/overused queue) could be reclaimed/preempted.

**Comment by Thor-wl** [comment]
> > 
> 
> I see your point. But for AI workload, gang is by default scheduling policy(Since most AI frameworks don't support failover). In such situation, reclaim will not work at all. So I think "pod number is less than minAvailable" should support configure, let user to decide if gang job(lower priority/overused queue) could be reclaimed/preempted.

I agree with your idea. That will be more flexible though it may improve a little bit learning cost for users. Can you give a design and PR for that?

**Comment by SoutherLea** [comment]
same issue. 
In my scenario, the low-priority job has only one pod, and I want the low-priority queue to be evicted for the portion that exceeds its weight. The minAvaliable of such tasks is 1, and in the current logic, such tasks cannot be reclaimed.

**Comment by nkflash** [comment]
> same issue.
> When scheduling batch v1 job with volcano. In my scenario, the low-priority job has only one pod, and I want the low-priority queue to be evicted for the portion that exceeds its weight. The minAvaliable of such tasks is 1, and in the current logic, such tasks cannot be reclaimed.

One work around:
- name: gang
enableJobStarving: false
enableReclaimable: false
enablePreemptable: false

**Comment by SoutherLea** [comment]
> 
I modified the logic by referring to the gang logic of kube-bath.
[https://github.com/kubernetes-sigs/kube-batch/blob/86b2ba217ad76340221fdc9b4ec16f3209dbff7e/pkg/scheduler/plugins/gang/gang.go](url)

```
if jobOccupiedMap[job.UID] > job.MinAvailable || job.MinAvailable == 1 {
	jobOccupiedMap[job.UID]--victims = append(victims, preemptee)
}
```
